encryption/src/main/java/org/apache/solr/encryption/EncryptionRequestHandler.java

```diff
@@ -260,12 +339,133 @@ public void handleRequestBody(SolrQueryRequest req, SolrQueryResponse rsp) throw
+        Collection<Callable<State>> encryptRequests = new ArrayList<>(slices.size());
+        final String collectionNameFinal = collectionName;
+        for (Slice slice : slices) {
+          Replica replica = slice.getLeader();
+          if (replica == null) {
+            log.error("No leader found for shard {}", slice.getName());
+            collectionState = State.ERROR;
+            continue;
+          }
+          encryptRequests.add(() -> sendEncryptionRequestWithRetry(replica, params, solrClient.getClient(), keyId, collectionNameFinal));
+        }
+        try {
+          List<Future<State>> responses = timeAllowedMs <= 0 ?
+              executor.invokeAll(encryptRequests)
+              : executor.invokeAll(encryptRequests, timeAllowedMs, TimeUnit.MILLISECONDS);
+          if (isTimeout(maxTimeNs)) {
+            log.warn("Timeout distributing encryption request for keyId={} collection={}", keyId, collectionName);
+            if (collectionState == null || State.TIMEOUT.priority > collectionState.priority) {
+              collectionState = State.TIMEOUT;
+            }
+          }
+          for (Future<State> response : responses) {
+            State state;
+            try {
+              state = response.get();
+            } catch (ExecutionException e) {
+              collectionState = State.ERROR;
+              break;
+            }
+            if (collectionState == null || state.priority > collectionState.priority) {
+              collectionState = state;
+            }
+          }
+        } catch (InterruptedException e) {
+          collectionState = State.INTERRUPTED;
+        }
+        success = collectionState == null || collectionState.isSuccess();
+      }
+    } finally {
+      String statusValue = success ? STATUS_SUCCESS : STATUS_FAILURE;
+      rsp.add(STATUS, statusValue);
+      rsp.addToLog(STATUS, statusValue);
+      if (collectionState != null) {
+        rsp.add(ENCRYPTION_STATE, collectionState.value);
+        rsp.addToLog(ENCRYPTION_STATE, collectionState.value);
+      }
+      if (log.isInfoEnabled()) {
+        long timeMs = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTimeNs);
+        log.info("Responding encryption distributed state={} success={} for keyId={} collection={} timeMs={}",
+            (collectionState == null ? null : collectionState.value), success, keyId, collectionName, timeMs);
+      }
+    }
+  }
+
+  private SolrClientHolder getHttpSolrClient(SolrQueryRequest req) {
+    CoreContainer coreContainer = req.getCore().getCoreContainer();
+    CloudSolrClient cloudSolrClient = coreContainer.getSolrClientCache().getCloudSolrClient(coreContainer.getZkController().getZkClient().getZkServerAddress());
+    if (cloudSolrClient instanceof CloudHttp2SolrClient) {
+      return new SolrClientHolder(((CloudHttp2SolrClient) cloudSolrClient).getHttpClient(), false);
+    }
+    return new SolrClientHolder(new Http2SolrClient.Builder().build(), true);
```

Comment:
I don't think this is needed, nor is SolrClientHolder.  Just use an existing client.
You can use `coreContainer.getUpdateShardHandler().getUpdateOnlyHttpClient()`.
It's mildly debatable if it's the perfect client since an encryption request is sorta kinda an "update" but it's really not a big deal.
You shouldn't create new clients (which was your fallback here) as it's tricky to do it correctly inside Solr as there are some instrumentation & authentication matters to attend to that UpdateShardHandler works out for you.